File: controllers/productController.js
import Product from '../models/Product.js';
import Discount from '../models/Discount.js';

// Fetch all products — supports ?search, ?material, ?region query params
export const getProducts = async (req, res, next) => {
    try {
        const { search, material, region } = req.query;
        const query = {};

        if (search) {
            query.$or = [
                { title: { $regex: search, $options: 'i' } },
                { originRegion: { $regex: search, $options: 'i' } }
            ];
        }
        if (material) query.materialsUsed = { $in: [new RegExp(material, 'i')] };
        if (region)   query.originRegion  = { $regex: region, $options: 'i' };

        const products = await Product.find(query).populate('artisanId', 'name email storeName').sort({ createdAt: -1 });

        const activeDiscount = await Discount.findOne({ isActive: true });
        
        const data = products.map(p => {
            const doc = p.toObject();
            doc.originalPrice = doc.price;
            if (activeDiscount) {
                const discountAmount = (doc.price * activeDiscount.percentage) / 100;
                doc.price = Math.max(0, doc.price - discountAmount);
                doc.discountBadge = `${activeDiscount.percentage}% OFF`;
            }
            return doc;
        });

        res.status(200).json({ success: true, data });
    } catch (error) {
        next(error);
    }
};

// Create a new product (Artisan / Admin)
export const createProduct = async (req, res, next) => {
    try {
        const { title, description, price, materialsUsed, originRegion, inventory, image } = req.body;
        const artisanId = req.user?.id || req.user?._id;

        if (!title || !price || !inventory) {
            return res.status(400).json({ success: false, message: "Title, price, and inventory are required." });
        }
        if (!artisanId) {
            return res.status(401).json({ success: false, message: "Not authorized to create product." });
        }

        const newProduct = await Product.create({
            artisanId, title, description, price,
            materialsUsed: Array.isArray(materialsUsed)
                ? materialsUsed
                : (materialsUsed ? materialsUsed.split(',').map(s => s.trim()).filter(Boolean) : []),
            originRegion, inventory,
            image: image || 'https://via.placeholder.com/400'
        });

        res.status(201).json({ success: true, message: "Product created successfully", data: newProduct });
    } catch (error) {
        next(error);
    }
};

// Update a product — Admin can update any; Artisan can only update their own
export const updateProduct = async (req, res, next) => {
    try {
        const product = await Product.findById(req.params.id);
        const requesterId = req.user?.id || req.user?._id;

        if (!product) return res.status(404).json({ success: false, message: "Product not found" });

        if (product.artisanId.toString() !== requesterId && req.user.role !== 'Admin') {
            return res.status(403).json({ success: false, message: "Not authorized to edit this product" });
        }

        const { title, price, inventory, originRegion, image, materialsUsed } = req.body;
        const updates = { title, price, inventory, originRegion, image };
        if (materialsUsed !== undefined) {
            updates.materialsUsed = Array.isArray(materialsUsed)
                ? materialsUsed
                : materialsUsed.split(',').map(s => s.trim()).filter(Boolean);
        }

        const updated = await Product.findByIdAndUpdate(req.params.id, updates, { returnDocument: 'after', runValidators: true });
        res.status(200).json({ success: true, message: "Product updated", data: updated });
    } catch (error) {
        next(error);
    }
};

// Delete a product — Admin or owning Artisan
export const deleteProduct = async (req, res, next) => {
    try {
        const product = await Product.findById(req.params.id);
        const requesterId = req.user?.id || req.user?._id;

        if (!product) return res.status(404).json({ success: false, message: "Product not found" });

        if (product.artisanId.toString() !== requesterId && req.user.role !== 'Admin') {
            return res.status(403).json({ success: false, message: "Not authorized to delete this asset" });
        }

        await Product.findByIdAndDelete(req.params.id);
        res.status(200).json({ success: true, message: "Asset removed from network" });
    } catch (error) {
        next(error);
    }
};

File: src/main/java/com/globalhandlooms/controller/ProductController.java
package com.globalhandlooms.controller;

import com.globalhandlooms.dto.ApiResponse;
import com.globalhandlooms.dto.product.CreateProductRequest;
import com.globalhandlooms.dto.product.UpdateProductRequest;
import com.globalhandlooms.model.Product;
import com.globalhandlooms.security.UserPrincipal;
import com.globalhandlooms.service.ProductService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/products")
@RequiredArgsConstructor
public class ProductController {

    private final ProductService productService;

    // GET /api/products?search=&material=&region=   (public)
    @GetMapping
    public ResponseEntity<ApiResponse> getProducts(
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String material,
            @RequestParam(required = false) String region) {
        List<Map<String, Object>> data = productService.getProducts(search, material, region);
        return ResponseEntity.ok(ApiResponse.builder().success(true).data(data).build());
    }

    // GET /api/products/artisan   (Artisan only — own listings)
    @GetMapping("/artisan")
    @PreAuthorize("hasAuthority('Artisan') or hasAuthority('Admin')")
    public ResponseEntity<?> getArtisanProducts(@AuthenticationPrincipal UserPrincipal user) {
        List<Product> data = productService.getArtisanProducts(user.id());
        return ResponseEntity.ok(Map.of("success", true, "data", data));
    }

    // POST /api/products   (Artisan or Admin)
    @PostMapping
    @PreAuthorize("hasAuthority('Artisan') or hasAuthority('Admin')")
    public ResponseEntity<ApiResponse> createProduct(
            @RequestBody CreateProductRequest req,
            @AuthenticationPrincipal UserPrincipal user) {
        Product created = productService.createProduct(user.id(), req);
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponse.builder().success(true).message("Product created successfully").data(created).build());
    }

    // PUT /api/products/:id   (Admin or owning Artisan)
    @PutMapping("/{id}")
    @PreAuthorize("hasAuthority('Admin') or hasAuthority('Artisan')")
    public ResponseEntity<ApiResponse> updateProduct(
            @PathVariable String id,
            @RequestBody UpdateProductRequest req,
            @AuthenticationPrincipal UserPrincipal user) {
        Product updated = productService.updateProduct(id, user.id(), user.role(), req);
        return ResponseEntity.ok(ApiResponse.builder().success(true).message("Product updated").data(updated).build());
    }

    // DELETE /api/products/:id   (Admin or owning Artisan)
    @DeleteMapping("/{id}")
    @PreAuthorize("hasAuthority('Admin') or hasAuthority('Artisan')")
    public ResponseEntity<ApiResponse> deleteProduct(
            @PathVariable String id,
            @AuthenticationPrincipal UserPrincipal user) {
        productService.deleteProduct(id, user.id(), user.role());
        return ResponseEntity.ok(ApiResponse.builder().success(true).message("Asset removed from network").build());
    }
}
